From 4247d7c5a0c9a5133245b935eef017149f49de87 Mon Sep 17 00:00:00 2001
From: Lars Knoll <lars.knoll@qt.io>
Date: Wed, 23 Jan 2019 14:12:47 +0100
Subject: [PATCH 147/166] Fix QTextTable:insertRows() for tables with spanning
 cells

Don't resize the height of cells spanning several columns
multiple times.

Fixes: QTBUG-36713
Change-Id: I5eb45892f2008e6a4f85745b56efd04323e25673
Reviewed-by: Eskil Abrahamsen Blomfeldt <eskil.abrahamsen-blomfeldt@qt.io>
---
 src/gui/text/qtexttable.cpp                       | 12 ++++++++----
 tests/auto/gui/text/qtexttable/tst_qtexttable.cpp |  6 ++++++
 2 files changed, 14 insertions(+), 4 deletions(-)

diff --git a/src/gui/text/qtexttable.cpp b/src/gui/text/qtexttable.cpp
index 9639c18d2b..39f26d5d42 100644
--- a/src/gui/text/qtexttable.cpp
+++ b/src/gui/text/qtexttable.cpp
@@ -696,18 +696,22 @@ void QTextTable::insertRows(int pos, int num)
     int extended = 0;
     int insert_before = 0;
     if (pos > 0 && pos < d->nRows) {
+        int lastCell = -1;
         for (int i = 0; i < d->nCols; ++i) {
             int cell = d->grid[pos*d->nCols + i];
             if (cell == d->grid[(pos-1)*d->nCols+i]) {
                 // cell spans the insertion place, extend it
-                QTextDocumentPrivate::FragmentIterator it(&p->fragmentMap(), cell);
-                QTextCharFormat fmt = c->charFormat(it->format);
-                fmt.setTableCellRowSpan(fmt.tableCellRowSpan() + num);
-                p->setCharFormat(it.position(), 1, fmt);
+                if (cell != lastCell) {
+                    QTextDocumentPrivate::FragmentIterator it(&p->fragmentMap(), cell);
+                    QTextCharFormat fmt = c->charFormat(it->format);
+                    fmt.setTableCellRowSpan(fmt.tableCellRowSpan() + num);
+                    p->setCharFormat(it.position(), 1, fmt);
+                }
                 extended++;
             } else if (!insert_before) {
                 insert_before = cell;
             }
+            lastCell = cell;
         }
     } else {
         insert_before = (pos == 0 ? d->grid[0] : d->fragment_end);
diff --git a/tests/auto/gui/text/qtexttable/tst_qtexttable.cpp b/tests/auto/gui/text/qtexttable/tst_qtexttable.cpp
index 29eef506bf..22f00c677d 100644
--- a/tests/auto/gui/text/qtexttable/tst_qtexttable.cpp
+++ b/tests/auto/gui/text/qtexttable/tst_qtexttable.cpp
@@ -431,6 +431,12 @@ void tst_QTextTable::insertRows()
     table->insertRows(5, 2);
     QCOMPARE(table->rows(), 7);
 
+    table = cursor.insertTable(5,5);
+    table->mergeCells(0,0,3,3);
+    table->insertRows(2,1);
+
+    QCOMPARE(table->rows(), 6);
+
 }
 
 void tst_QTextTable::deleteInTable()
-- 
2.20.1

